### PR TITLE
Class method support, usage in the browser

### PR DESCRIPTION
**Pull request by jneuendorf**
### 1. class method support

Class methods should now behave like instance methods.
Thus CoffeeScript's native `super` can be used like so:

```coffee
class A
    @classMethod: () ->
        return super() + 'Base1'

class B
    @classMethod: () ->
        return 'Base2'

class C extends multi A, B

# C.classMethod() is 'Base2Base1'
```

### 2. usage in the browser

`make` now also creates a `lib/browser.heterarchy.js` file that can be used in the browser.
A description is available in the `README`.
`make` uses a python script to modify `heterarchy.litcoffee` before compiling.
I thought of python due to "CoffeeScript, à-la Python" but maybe a JavaScript/CoffeeScript file run with `node` would fit better because heterarchy is a node js module...

### CODE REVIEW

**Comment by arximboldi** [comment]
Hi @jneuendorf!

Thanks a lot for this PR! It looks like really usefull stuff. I will take a deeper look at it tomorrow. 

On the topic of Python vs Node.js: yes, I would prefer if that script was converted to Node.js if you have time for that. I like Python, but I'd rather not add another external dependency to the project.

**Comment by arximboldi** [comment]
Btw, it seems like the commits are not associated to your GitHub account. This is the because the @gmx.de account you are using for the commits is not associated to your GitHub account. You can do that in `settings > emails` in the GitHub page. It's not important, but just thought you might want to do that to get the credit in your profile timeline once it is merged :smile:

**Comment by jneuendorf** [comment]
@arximboldi I converted the script according to https://github.com/arximboldi/heterarchy/pull/5#issuecomment-279096778 
The script apparently does not work on windows (because of the `/dev/stdin`) - but I guess that's alright since the whole `make` process is also unix specific, right?

File: heterarchy.litcoffee [hunk 1]
```diff
@@ -12,9 +12,9 @@ Dylan paper](http://192.220.96.201/dylan/linearization-oopsla96.html).
     {head, tail, map, find, some, without, isEmpty, every, memoize, reject,
      partial, isEqual, reduce} = require 'underscore'
 
-    assert = (value, error) ->
+    assert = (value, errorMessage) ->
         if not value
-            throw new Error(if error? then error else "Assertion failed")
+            throw new Error(errorMessage)
```
Comment (arximboldi) [inline heterarchy.litcoffee:44]: Why did you change this? The way it was before, you could omit the second argument.

File: README.md [hunk 2]
```diff
@@ -68,6 +68,48 @@ install the library with:
...
 > npm install heterarchy
 
+### Usage in the browser
+
+After cloning or downloading the repository you can build all necessary files with
+
+> make
+
+This will create `browser.heterarchy.js` (and `heterarchy.js`) in the `lib/` folder.
+Including that file in the browser will result in a global `heterarchy` variable that contains the normally exported functions.
```
Comment (arximboldi) [inline README.md]: I try to keep README.md lines at 72 characters.

File: heterarchy.litcoffee [hunk 3]
```diff
@@ -63,7 +63,18 @@ behaves like a class that would be have such a hierarchy.
             class Result extends generate tail linearization
                 __mro__: linearization
                 constructor: reparent next, @, next::constructor
+                # FILL UP MISSING CLASS ATTRIBUTES
+                # copy class attributes from `next` to `this` that only exist on `next`
                 copyOwn next, @
+                # ADJUST CLASS METHOD (so the MRO is used)
+                # copy class methods from `next` to `this` that are implemented by `next`
+                # note: this is basically the same as copyOwn but ignoring `if not to.hasOwnProperty key`
+                # This is due to the fact that when inheriting all class attributes are copied to `this` by CoffeeScript.
```
Comment (arximboldi) [inline heterarchy.litcoffee]: I keep code lines at 80 characters.

File: heterarchy.litcoffee [hunk 4]
```diff
@@ -92,13 +103,13 @@ it in the later.
             value
         else if value is oldklass::constructor and inherited(oldklass) is Object
             superctor = inherited(newklass)::constructor
-            ->
+            () ->
```
Comment (arximboldi) [inline heterarchy.litcoffee]: Is this changed because of a linter?

File: script/browserify.js [hunk 5]
```diff
@@ -0,0 +1,11 @@
+#!/usr/bin/env node
```
Comment (arximboldi) [inline script/browserify.js]: The script name is a bit confusing. There is already a tool called `browserify` (http://browserify.org/). 
In fact, I wonder... why do you need this script, could you not just use the `browserify` tool externally to build a browser ready package? That tool bundles your script together with the dependencies like `heterarchy` and `underscore` automatically.

File: test/heterarchy.spec.coffee [hunk 6]
```diff
@@ -141,20 +156,54 @@ describe 'heterarchy', ->
...
+                            return super() + 'Base1'
+
+                    class B
+                        @classMethod: () ->
+                            return 'Base2'
+
+                    class C extends multi A, B
+
+                    C.classMethod().should.equal 'Base2Base1'
+                )(0, 0, 0)
```
Comment (arximboldi) [inline test/heterarchy.spec.coffee]: Shouldn't this also work without passing the 0-s, (the values just get `undefined`)?

File: test/heterarchy.spec.coffee [hunk 7]
```diff
@@ -165,25 +214,31 @@ describe 'heterarchy', ->
...
             obj.deriv .should.equal 'deriv'
 
         it 'allows access class properties', ->
-            obj = new Deriv
+            obj = new Deriv()
             obj.classProperty .should.equal 42
 
         it 'allows class properties to be set via object', ->
-            obj = new Deriv
+            obj = new Deriv()
```
Comment (arximboldi) [inline test/heterarchy.spec.coffee]: Again, why add parens all over? 😄

File: heterarchy.litcoffee [hunk 8]
```diff
@@ -63,7 +63,18 @@ behaves like a class that would be have such a hierarchy.
...
                 constructor: reparent next, @, next::constructor
+                # FILL UP MISSING CLASS ATTRIBUTES
+                # copy class attributes from `next` to `this` that only exist on `next`
                 copyOwn next, @
+                # ADJUST CLASS METHOD (so the MRO is used)
+                # copy class methods from `next` to `this` that are implemented by `next`
+                # note: this is basically the same as copyOwn but ignoring `if not to.hasOwnProperty key`
+                # This is due to the fact that when inheriting all class attributes are copied to `this` by CoffeeScript.
+                for own key, value of next when value instanceof Function
+                    @[key] = partial(reparent, next, @)(value)
```
Comment (arximboldi) [inline heterarchy.litcoffee]: Note that the `reparent` function already checks if `vale instanceof Function`. This means that this loop can be merged with the previous one. I guess this means that the `copyOwn` function would only be called once, and it would not make sense. We could just replace the two calls to `copyOwn` and your loop by something like this:
```coffee
                for own key, value of next
                    @[key] = reparent next, @, value
                for own key, value of next::
                    if not @::.hasOwnProperty key
                        @::[key] = reparent next, @, value
```

File: test/heterarchy.spec.coffee [hunk 9]
```diff
@@ -58,42 +58,56 @@ describe 'heterarchy', ->
...
         method: -> "F>#{super}"
+        @method: -> "F>#{super}"
+        @overrideNoSuper: () -> "f"
 
     class G extends multi D, F
         constructor: ->
             super
             @g = 'g'
         method: -> "G>#{super}"
+        @method: -> "G>#{super}"
```
Comment (arximboldi) [inline test/heterarchy.spec.coffee]: It would be nice to give the class method a different name, to make sure the methods checking instance methods are not passing because of the override.

**Review by arximboldi — COMMENTED** [review]
Hey!

Sorry it took me so long to review this... I know I promised I would do this a week ago. I have added lots of comments/questiosn, a lot of it is nitpicking, so you don't need to address them... I could change it myself after merging it. just write a comment telling me when you think you are done or if you prefer to do you further changes.

In general, I really like the change... this class method issue was an important missing feature! You also improved the docs and fixed lots of typos... I am such a clumsy writer 😄 

A couple of things that could me improved for another submissions:

- I prefer when each commit has a specific purpose, and makes the tests pass. You have a lot of back-and-forth in the branch, so I could not review per-commit but do a whole review.
- In this regard, "experimental" stuff can be squashed together.
- There is a commit that says "added linter file", but no linter file is added to the repository.
- You changed a lot of stylistic things, I guess, suggested by your linter. 
- In general it is good practice to keep the style consistent with what is already there. For example, I like avoiding parenthesis--you changed it in some places, and now the style is mixed. Same goes to line lenth.
- Part of that is my fault, because I had no link to a style guide, sorry about that.

Once again, thanks a lot!

File: heterarchy.litcoffee (hunk 1 above)
Comment (jneuendorf) [reply heterarchy.litcoffee:44]: Yes, usability-wise it was better (or at least more generic). I changed this, because the (I believe) `error?` was always true. Thus, `test-coverage` said one code branch was never executed.
Maybe the better way is to add an assertion call to the test spec. 😉

File: heterarchy.litcoffee (hunk 3 above)
Comment (jneuendorf) [reply heterarchy.litcoffee]: Yep, python-like 😄 Depending on how complicated squashing is, that could be one of those things you could do after the merge (otherwise it would just create another commit that I'd need to handle/squash)

File: README.md (hunk 2 above)
Comment (jneuendorf) [reply README.md]: see https://github.com/arximboldi/heterarchy/pull/5#discussion_r101874284

File: script/browserify.js (hunk 5 above)
Comment (jneuendorf) [reply script/browserify.js]: I know about browserify and that's where the name came from. For some reason, the real browserify did not always work for me (lately) and since I am using heterarchy at work in the browser (where I could not get it to work with browserify) I wrote the little script.

I think the underscore lib writes the javascript in a way that browserification is not even necessary. Maybe that would be the best way. The question then would be how to define the namespaces.
I just recently created a repo called [JsUml](https://github.com/jneuendorf/JsUml) where it tried to do it like underscore (see https://github.com/jneuendorf/JsUml/blob/master/js_uml.coffee).

File: heterarchy.litcoffee (hunk 4 above)
Comment (jneuendorf) [reply heterarchy.litcoffee]: Sorry about the parens ^^ For function and constructors I **personally** dislike not writing them. But this that's your repo I would change 'em back 😉

File: test/heterarchy.spec.coffee (hunk 7 above)
Comment (jneuendorf) [reply test/heterarchy.spec.coffee]: see https://github.com/arximboldi/heterarchy/pull/5#discussion_r101874971

File: test/heterarchy.spec.coffee (hunk 6 above)
Comment (jneuendorf) [reply test/heterarchy.spec.coffee]: Yeah probably. I just didn't wanna use `do (A,B,C)` because that wouldn't work, right?

File: heterarchy.litcoffee (hunk 8 above)
Comment (jneuendorf) [reply heterarchy.litcoffee]: I'll look at this later. But I think that's not quite true because of the `partial` and `reparent` functions that are differently used in the loops

File: test/heterarchy.spec.coffee (hunk 9 above)
Comment (jneuendorf) [reply test/heterarchy.spec.coffee]: Yeah, good point

File: script/browserify.js (hunk 5 above)
Comment (arximboldi) [reply script/browserify.js]: Oh, no problem. I am happy to add the feature if you find it useful for your project. I like the way you do it in that `js_uml` library, so feel free to do it that way here too if you want!

File: test/heterarchy.spec.coffee (hunk 6 above)
Comment (arximboldi) [reply test/heterarchy.spec.coffee]: Mmm, good point, actually I remember using `do -> ...` to create scopes in other project (Mixco?) which I find more elegant than the parens over the whole function...

File: heterarchy.litcoffee (hunk 3 above)
Comment (arximboldi) [reply heterarchy.litcoffee]: Cool, don't worry about it!

File: heterarchy.litcoffee (hunk 1 above)
Comment (arximboldi) [reply heterarchy.litcoffee:44]: Oh I see, as you wish then :smile:

File: test/heterarchy.spec.coffee (hunk 6 above)
Comment (jneuendorf) [reply test/heterarchy.spec.coffee]: It actually does work with

```coffee
class A
  @a = 1

do (A) ->
  class A
    @a = 2

alert A.a # this alerts 1
```

Because the local `A` is being assigned. Thus it does not matter if the local `A` equals the outer-scope `A` - tricky ^^
Thus `do (A,B,C)` should be ok

**Comment by jneuendorf** [comment]
@arximboldi

1. Concerning the coffee-linter file: I apparently screwed up this one. Would you like such a file (that matches your coding style)? It could look something like this:

```json
{
    "indentation" : {
        "level" : "warn",
        "value" : 4
    },
    "line_endings" : {
        "value" : "unix",
        "level" : "error"
    },
    "max_line_length": {
        "value": 80,
        "level": "error"
    },
    "arrow_spacing": {
        "level": "error"
    },
    "no_this": {
        "level": "error"
    },
    "prefer_english_operator": {
        "level": "error"
    },
    "space_operators" : {
        "level": "warn"
    },
    "spacing_after_comma": {
        "level": "error"
    }
}
```

2. I'll try to squash some commits once we're ready to merge 😉

File: test/heterarchy.spec.coffee (hunk 7 above)
Comment (jneuendorf) [reply test/heterarchy.spec.coffee]: fixed in 6374cd8

File: test/heterarchy.spec.coffee (hunk 6 above)
Comment (jneuendorf) [reply test/heterarchy.spec.coffee]: fixed in f1b2d0d

File: test/heterarchy.spec.coffee (hunk 9 above)
Comment (jneuendorf) [reply test/heterarchy.spec.coffee]: fixed in 2a9bfe8

File: heterarchy.litcoffee (hunk 8 above)
Comment (arximboldi) [reply heterarchy.litcoffee]: I tried it before commenting and it seemed to work though 😉

**Comment by arximboldi** [comment]
Cool, thanks a lot for all this effort. W.r.t. the linter file, that looks good to me. If you want, you can also add a Travis check for that.

Also, I think this contribution deserves mention in the `README`. I will add a contributors section for this, do you prefer to be mentioned by full name/github handle/email/link, or any combination of these?

File: heterarchy.litcoffee (hunk 8 above)
Comment (jneuendorf) [reply heterarchy.litcoffee]: Me, too; and the tests pass. I inlined the `copyOwn` function and got rid of the extra check for `value instanceof Function` to see how close that would be to your suggestion. Doing so resulted in almost the same code as yours. The only difference is that `partial` function is used.
My main question is: Why was `partial` used in the first place? The following 2 lines should be equal, right?

```coffee
partial(reparent, next, @) value
reparent next, @, value
```
If so, using `partial` just never made if I'm not mistaken but you must have used it for _some_ reason. 😉

Comment (arximboldi) [reply heterarchy.litcoffee]: Those two lines are equivalent.

I used `partial` to *curry* [1] the function passed to `copyOwn`, (`copyOwn` takes a function that *projects* the copied values, similar to `Array.map`). This is, we use `partial` to create a new function that binds `reparent`, `next`, and `@` to `reparent`, but leaves the last `value` argument free, thus creating a unary (with one argument) function out of a function with more arguments. These two expressions are equivalent:
```coffee
copyOwn next::, @::, partial reparent, next, @
copyOwn next::, @::, (v) -> reparent, next, @, v
```
i used `partial` because I was used to it from other languages tha support this pattern (it is also in Python `functools` btw.).

Of course, writing something like this...
```coffee
partial(reparent, next, @) value
```
...works but it is redundant, because it creates a new closure but then immediatelly applies it... we could just apply it to begin with using `reparent next, @, value` as you suggest. I think you added that line of code when inlining the `copyOwn` function, but forgot to also *inline* the partial call 😄 

[1] https://en.wikipedia.org/wiki/Currying

**Comment by jneuendorf** [comment]
Concerning the `readme`: I guess the full name and the github account would be cool 😄 

I just made heterarchy support node and the browser. But with that the coverage decreased because I could not get `test-coverage` to work with `mocha-phantomjs`. 😢  Do you have an idea? The `test`s themselves work though...
And shouldn't the tests also be run from `Travis CI`?

**Edit:** I just found [this hook](https://github.com/willembult/mocha-phantomjs-istanbul)

**Comment by arximboldi** [comment]
About running the tests on Travis: yes that would be cool, otherwise I might break this feature in the future by mistake...

On the coverage: I'd say don't worry if you don't find a solution. I am not so interested in coverage absolute numbers, but more in the deltas and per line analysis (to check if I made the wrong assumptions about some code/test).

File: heterarchy.litcoffee (hunk 8 above)
Comment (jneuendorf) [reply heterarchy.litcoffee]: Right, I inlined it and then was not 100% sure about why `partial`/currying was used and wanted another opinion. 😉 
Good to know that everything's fine now 😄

**Comment by jneuendorf** [comment]
Alright then, I think we're done here (I'd at least postpone the _coverage_ topic to have the new feature available).

This leaves me with the `git squash`. Can I just do the `rebase` in my repo and everything will work? Do you know any good tutorial? I've never done that.
I guess it would make sense to combine the commits into units (as far as that's possible), right?

**Comment by arximboldi** [comment]
@jneuendorf Yeah, the idea is to do a rebase interactive `git rebase -i HEAD^{number of commits back in the history you want to edit}`... admittedly I don't really know any good tutorial... I think I learnt this from coworkers back in the day. And admittedly, I used to hate it in the beginning and often messed up my branches, so I don't want to inflict this pain into you haha! I will merge it as it is if want.

**Comment by jneuendorf** [comment]
If _interactive rebase_ could mess up the work I think it would be best (for now) to merge everything the way it is.
I could then play around with rebasing locally or so 😉

**Comment by arximboldi** [comment]
Cool, merged!

**Comment by arximboldi** [comment]
Thanks a lot @jneuendorf
